crates/subspace-farmer/src/farmer_cache.rs

```diff
@@ -43,16 +48,83 @@ const IS_PIECE_MAYBE_STORED_TIMEOUT: Duration = Duration::from_millis(100);
 type HandlerFn<A> = Arc<dyn Fn(&A) + Send + Sync + 'static>;
 type Handler<A> = Bag<HandlerFn<A>, A>;
 
+#[derive(Debug, Clone)]
+struct FarmerCacheOffset(u32);
+
+impl FarmerCacheOffset {
+    fn new(cache_index: u8, cache_offset: u32) -> Option<Self> {
+        // Check if cache offset is valid
+        if cache_offset > CACHE_OFFSET_MAX {
+            warn!(
+                %cache_offset,
+                "Each cache has a maximum offset of {CACHE_OFFSET_MAX}; \
+                Any portion exceeding this limit will be ignored"
+            );
+            return None;
+        }
```

Comment:
This is an internal data structure, isn't there a way to guarantee that this will never happen? If so then we can just use `unreachable!()` here with explanation why code is not reachable. It might also be fine to just go with cache index and cache offset as separate fields for now to avoid this accidental complexity, will make many things much easier.

Reply:
I see what you mean, simplifying the code would have been a better choice. There's not much to change here, so let me take care of that.